File: css-duplicate-checker.py
#!/usr/bin/env python3
"""
CSS Duplicate Checker - Python Version
Automatically detects duplicate CSS rules and selectors
"""

import re
import json
import sys
from collections import defaultdict
from datetime import datetime

class CSSDuplicateChecker:
    def __init__(self, css_file_path):
        self.css_file_path = css_file_path
        self.css_content = ""
        self.duplicates = {
            'selectors': defaultdict(list),
            'rules': defaultdict(list),
            'properties': defaultdict(list)
        }

    def read_css_file(self):
        """Read and parse CSS file"""
        try:
            with open(self.css_file_path, 'r', encoding='utf-8') as file:
                self.css_content = file.read()
            file_size = len(self.css_content) / 1024
            print(f"✅ Successfully read CSS file: {self.css_file_path}")
            print(f"📊 File size: {file_size:.2f} KB")
            return True
        except FileNotFoundError:
            print(f"❌ CSS file not found: {self.css_file_path}")
            return False
        except Exception as e:
            print(f"❌ Error reading CSS file: {e}")
            return False

    def parse_css(self):
        """Extract CSS rules and selectors"""
        rules = []
        lines = self.css_content.split('\n')
        current_rule = None
        brace_count = 0
        rule_start_line = 0

        for i, line in enumerate(lines):
            line = line.strip()
            
            # Skip comments and empty lines
            if line.startswith('/*') or line.startswith('*') or line == '':
                continue

            # Check for selector (contains { but not inside a rule)
            if '{' in line and brace_count == 0:
                if current_rule:
                    rules.append(current_rule)
                
                selector = line.split('{')[0].strip()
                current_rule = {
                    'selector': selector,
                    'properties': [],
                    'start_line': i + 1,
                    'end_line': i + 1,
                    'full_text': line
                }
                brace_count = 1
                rule_start_line = i + 1
            # Inside a rule
            elif current_rule and brace_count > 0:
                current_rule['full_text'] += '\n' + line
                current_rule['end_line'] = i + 1

                # Count braces
                brace_count += line.count('{') - line.count('}')

                # Rule ends
                if brace_count == 0:
                    rules.append(current_rule)
                    current_rule = None
                # Property inside rule
                elif ':' in line and '{' not in line and '}' not in line:
                    parts = line.split(':', 1)
                    property_name = parts[0].strip()
                    value = parts[1].strip().rstrip(';')
                    current_rule['properties'].append({
                        'property': property_name,
                        'value': value,
                        'line': i + 1
                    })

        # Add last rule if exists
        if current_rule:
            rules.append(current_rule)

        return rules

    def find_duplicate_selectors(self, rules):
        """Find duplicate selectors"""
        selector_map = defaultdict(list)
        
        for rule in rules:
            selector_map[rule['selector']].append(rule)

        # Filter only duplicates
        for selector, rule_list in selector_map.items():
            if len(rule_list) > 1:
                self.duplicates['selectors'][selector] = rule_list

    def find_similar_rules(self, rules):
        """Find similar rules (same selector with different properties)"""
        selector_map = defaultdict(list)
        
        for rule in rules:
            selector_map[rule['selector']].append(rule)

        # Filter selectors with multiple rule definitions
        for selector, rule_list in selector_map.items():
            if len(rule_list) > 1:
                self.duplicates['rules'][selector] = rule_list

    def find_duplicate_properties(self, rules):
        """Find duplicate property combinations"""
        property_map = defaultdict(list)
        
        for rule in rules:
            if not rule['properties']:
                continue
            
            # Create a signature of all properties
            signature = '|'.join(sorted([f"{p['property']}:{p['value']}" for p in rule['properties']]))
            
            property_map[signature].append(rule)

        # Filter only duplicates
        for signature, rule_list in property_map.items():
            if len(rule_list) > 1:
                self.duplicates['properties'][signature] = rule_list

    def generate_report(self):
        """Generate detailed report"""
        print('\n🔍 CSS DUPLICATE ANALYSIS REPORT')
        print('=' * 50)

        # Duplicate Selectors Report
        if self.duplicates['selectors']:
            print(f'\n❌ DUPLICATE SELECTORS ({len(self.duplicates["selectors"])} found):')
            print('-' * 30)
            
            for selector, rules in self.duplicates['selectors'].items():
                print(f'\n📌 Selector: {selector}')
                print(f'   Found {len(rules)} identical rules:')
                for i, rule in enumerate(rules):
                    print(f'   {i + 1}. Lines {rule["start_line"]}-{rule["end_line"]}')
        else:
            print('\n✅ No duplicate selectors found!')

        # Similar Rules Report
        if self.duplicates['rules']:
            print(f'\n⚠️  SIMILAR RULES ({len(self.duplicates["rules"])} found):')
            print('-' * 30)
            
            for selector, rules in self.duplicates['rules'].items():
                print(f'\n📌 Selector: {selector}')
                print(f'   Found {len(rules)} different rule definitions:')
                for i, rule in enumerate(rules):
                    print(f'   {i + 1}. Lines {rule["start_line"]}-{rule["end_line"]}')
                    print(f'      Properties: {len(rule["properties"])}')
        else:
            print('\n✅ No similar rules found!')

        # Duplicate Properties Report
        if self.duplicates['properties']:
            print(f'\n🔄 DUPLICATE PROPERTY COMBINATIONS ({len(self.duplicates["properties"])} found):')
            print('-' * 30)
            
            for signature, rules in self.duplicates['properties'].items():
                display_signature = signature[:100] + '...' if len(signature) > 100 else signature
                print(f'\n📌 Properties: {display_signature}')
                print(f'   Found in {len(rules)} different selectors:')
                for i, rule in enumerate(rules):
                    print(f'   {i + 1}. {rule["selector"]} (Lines {rule["start_line"]}-{rule["end_line"]})')
        else:
            print('\n✅ No duplicate property combinations found!')

        # Summary
        total_duplicates = (len(self.duplicates['selectors']) + 
                           len(self.duplicates['rules']) + 
                           len(self.duplicates['properties']))
        
        print('\n📊 SUMMARY:')
        print('-' * 30)
        print(f'Total duplicate issues found: {total_duplicates}')
        print(f'Duplicate selectors: {len(self.duplicates["selectors"])}')
        print(f'Similar rules: {len(self.duplicates["rules"])}')
        print(f'Duplicate properties: {len(self.duplicates["properties"])}')
        
        if total_duplicates == 0:
            print('\n🎉 Congratulations! Your CSS file is clean of duplicates!')
        else:
            print('\n💡 Consider consolidating these duplicates to improve maintainability.')

    def export_results(self, output_file):
        """Export results to JSON"""
        results = {
            'timestamp': datetime.now().isoformat(),
            'file': self.css_file_path,
            'summary': {
                'total_duplicates': (len(self.duplicates['selectors']) + 
                                   len(self.duplicates['rules']) + 
                                   len(self.duplicates['properties'])),
                'duplicate_selectors': len(self.duplicates['selectors']),
                'similar_rules': len(self.duplicates['rules']),
                'duplicate_properties': len(self.duplicates['properties'])
            },
            'duplicates': {
                'selectors': [
                    {
                        'selector': selector,
                        'count': len(rules),
                        'locations': [{'start_line': r['start_line'], 'end_line': r['end_line']} for r in rules]
                    }
                    for selector, rules in self.duplicates['selectors'].items()
                ],
                'rules': [
                    {
                        'selector': selector,
                        'count': len(rules),
                        'locations': [{'start_line': r['start_line'], 'end_line': r['end_line'], 'property_count': len(r['properties'])} for r in rules]
                    }
                    for selector, rules in self.duplicates['rules'].items()
                ],
                'properties': [
                    {
                        'signature': signature[:200],
                        'count': len(rules),
                        'selectors': [{'selector': r['selector'], 'start_line': r['start_line'], 'end_line': r['end_line']} for r in rules]
                    }
                    for signature, rules in self.duplicates['properties'].items()
                ]
            }
        }

        try:
            with open(output_file, 'w', encoding='utf-8') as file:
                json.dump(results, file, indent=2, ensure_ascii=False)
            print(f'\n📄 Results exported to: {output_file}')
        except Exception as e:
            print(f'\n❌ Error exporting results: {e}')

    def analyze(self):
        """Main analysis method"""
        print('🚀 Starting CSS Duplicate Analysis...\n')
        
        if not self.read_css_file():
            return False

        rules = self.parse_css()
        print(f'📋 Parsed {len(rules)} CSS rules')

        self.find_duplicate_selectors(rules)
        self.find_similar_rules(rules)
        self.find_duplicate_properties(rules)

        self.generate_report()
        self.export_results('css-duplicates-report.json')

        return True

def main():
    css_file = sys.argv[1] if len(sys.argv) > 1 else 'frontend/src/styles/globals.css'
    
    checker = CSSDuplicateChecker(css_file)
    checker.analyze()

if __name__ == '__main__':
    main()

File: css-duplicate-checker.js
#!/usr/bin/env node

/**
 * CSS Duplicate Checker
 * Automatically detects duplicate CSS rules and selectors
 */

const fs = require('fs');
const path = require('path');

class CSSDuplicateChecker {
  constructor(cssFilePath) {
    this.cssFilePath = cssFilePath;
    this.cssContent = '';
    this.duplicates = {
      selectors: new Map(),
      rules: new Map(),
      properties: new Map()
    };
  }

  // Read and parse CSS file
  readCSSFile() {
    try {
      this.cssContent = fs.readFileSync(this.cssFilePath, 'utf8');
      console.log(`✅ Successfully read CSS file: ${this.cssFilePath}`);
      console.log(`📊 File size: ${(this.cssContent.length / 1024).toFixed(2)} KB`);
      return true;
    } catch (error) {
      console.error(`❌ Error reading CSS file: ${error.message}`);
      return false;
    }
  }

  // Extract CSS rules and selectors
  parseCSS() {
    const rules = [];
    const lines = this.cssContent.split('\n');
    let currentRule = null;
    let braceCount = 0;
    let ruleStartLine = 0;

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i].trim();
      
      // Skip comments and empty lines
      if (line.startsWith('/*') || line.startsWith('*') || line === '') {
        continue;
      }

      // Check for selector (contains { but not inside a rule)
      if (line.includes('{') && braceCount === 0) {
        if (currentRule) {
          rules.push(currentRule);
        }
        
        const selector = line.split('{')[0].trim();
        currentRule = {
          selector: selector,
          properties: [],
          startLine: i + 1,
          endLine: i + 1,
          fullText: line
        };
        braceCount = 1;
        ruleStartLine = i + 1;
      }
      // Inside a rule
      else if (currentRule && braceCount > 0) {
        currentRule.fullText += '\n' + line;
        currentRule.endLine = i + 1;

        // Count braces
        braceCount += (line.match(/\{/g) || []).length;
        braceCount -= (line.match(/\}/g) || []).length;

        // Rule ends
        if (braceCount === 0) {
          rules.push(currentRule);
          currentRule = null;
        }
        // Property inside rule
        else if (line.includes(':') && !line.includes('{') && !line.includes('}')) {
          const property = line.split(':')[0].trim();
          const value = line.split(':').slice(1).join(':').trim().replace(';', '');
          currentRule.properties.push({ property, value, line: i + 1 });
        }
      }
    }

    // Add last rule if exists
    if (currentRule) {
      rules.push(currentRule);
    }

    return rules;
  }

  // Find duplicate selectors
  findDuplicateSelectors(rules) {
    const selectorMap = new Map();
    
    rules.forEach(rule => {
      const selector = rule.selector;
      if (selectorMap.has(selector)) {
        selectorMap.get(selector).push(rule);
      } else {
        selectorMap.set(selector, [rule]);
      }
    });

    // Filter only duplicates
    selectorMap.forEach((rules, selector) => {
      if (rules.length > 1) {
        this.duplicates.selectors.set(selector, rules);
      }
    });
  }

  // Find duplicate property combinations
  findDuplicateProperties(rules) {
    const propertyMap = new Map();
    
    rules.forEach(rule => {
      if (rule.properties.length === 0) return;
      
      // Create a signature of all properties
      const signature = rule.properties
        .map(p => `${p.property}:${p.value}`)
        .sort()
        .join('|');
      
      if (propertyMap.has(signature)) {
        propertyMap.get(signature).push(rule);
      } else {
        propertyMap.set(signature, [rule]);
      }
    });

    // Filter only duplicates
    propertyMap.forEach((rules, signature) => {
      if (rules.length > 1) {
        this.duplicates.properties.set(signature, rules);
      }
    });
  }

  // Find similar rules (same selector with different properties)
  findSimilarRules(rules) {
    const similarMap = new Map();
    
    rules.forEach(rule => {
      const selector = rule.selector;
      if (similarMap.has(selector)) {
        similarMap.get(selector).push(rule);
      } else {
        similarMap.set(selector, [rule]);
      }
    });

    // Filter selectors with multiple rule definitions
    similarMap.forEach((rules, selector) => {
      if (rules.length > 1) {
        this.duplicates.rules.set(selector, rules);
      }
    });
  }

  // Generate detailed report
  generateReport() {
    console.log('\n🔍 CSS DUPLICATE ANALYSIS REPORT');
    console.log('=' .repeat(50));

    // Duplicate Selectors Report
    if (this.duplicates.selectors.size > 0) {
      console.log(`\n❌ DUPLICATE SELECTORS (${this.duplicates.selectors.size} found):`);
      console.log('-'.repeat(30));
      
      this.duplicates.selectors.forEach((rules, selector) => {
        console.log(`\n📌 Selector: ${selector}`);
        console.log(`   Found ${rules.length} identical rules:`);
        rules.forEach((rule, index) => {
          console.log(`   ${index + 1}. Lines ${rule.startLine}-${rule.endLine}`);
        });
      });
    } else {
      console.log('\n✅ No duplicate selectors found!');
    }

    // Similar Rules Report
    if (this.duplicates.rules.size > 0) {
      console.log(`\n⚠️  SIMILAR RULES (${this.duplicates.rules.size} found):`);
      console.log('-'.repeat(30));
      
      this.duplicates.rules.forEach((rules, selector) => {
        console.log(`\n📌 Selector: ${selector}`);
        console.log(`   Found ${rules.length} different rule definitions:`);
        rules.forEach((rule, index) => {
          console.log(`   ${index + 1}. Lines ${rule.startLine}-${rule.endLine}`);
          console.log(`      Properties: ${rule.properties.length}`);
        });
      });
    } else {
      console.log('\n✅ No similar rules found!');
    }

    // Duplicate Properties Report
    if (this.duplicates.properties.size > 0) {
      console.log(`\n🔄 DUPLICATE PROPERTY COMBINATIONS (${this.duplicates.properties.size} found):`);
      console.log('-'.repeat(30));
      
      this.duplicates.properties.forEach((rules, signature) => {
        console.log(`\n📌 Properties: ${signature.substring(0, 100)}${signature.length > 100 ? '...' : ''}`);
        console.log(`   Found in ${rules.length} different selectors:`);
        rules.forEach((rule, index) => {
          console.log(`   ${index + 1}. ${rule.selector} (Lines ${rule.startLine}-${rule.endLine})`);
        });
      });
    } else {
      console.log('\n✅ No duplicate property combinations found!');
    }

    // Summary
    const totalDuplicates = this.duplicates.selectors.size + 
                           this.duplicates.rules.size + 
                           this.duplicates.properties.size;
    
    console.log('\n📊 SUMMARY:');
    console.log('-'.repeat(30));
    console.log(`Total duplicate issues found: ${totalDuplicates}`);
    console.log(`Duplicate selectors: ${this.duplicates.selectors.size}`);
    console.log(`Similar rules: ${this.duplicates.rules.size}`);
    console.log(`Duplicate properties: ${this.duplicates.properties.size}`);
    
    if (totalDuplicates === 0) {
      console.log('\n🎉 Congratulations! Your CSS file is clean of duplicates!');
    } else {
      console.log('\n💡 Consider consolidating these duplicates to improve maintainability.');
    }
  }

  // Export results to JSON
  exportResults(outputFile) {
    const results = {
      timestamp: new Date().toISOString(),
      file: this.cssFilePath,
      summary: {
        totalDuplicates: this.duplicates.selectors.size + this.duplicates.rules.size + this.duplicates.properties.size,
        duplicateSelectors: this.duplicates.selectors.size,
        similarRules: this.duplicates.rules.size,
        duplicateProperties: this.duplicates.properties.size
      },
      duplicates: {
        selectors: Array.from(this.duplicates.selectors.entries()).map(([selector, rules]) => ({
          selector,
          count: rules.length,
          locations: rules.map(rule => ({ startLine: rule.startLine, endLine: rule.endLine }))
        })),
        rules: Array.from(this.duplicates.rules.entries()).map(([selector, rules]) => ({
          selector,
          count: rules.length,
          locations: rules.map(rule => ({ startLine: rule.startLine, endLine: rule.endLine, propertyCount: rule.properties.length }))
        })),
        properties: Array.from(this.duplicates.properties.entries()).map(([signature, rules]) => ({
          signature: signature.substring(0, 200),
          count: rules.length,
          selectors: rules.map(rule => ({ selector: rule.selector, startLine: rule.startLine, endLine: rule.endLine }))
        }))
      }
    };

    try {
      fs.writeFileSync(outputFile, JSON.stringify(results, null, 2));
      console.log(`\n📄 Results exported to: ${outputFile}`);
    } catch (error) {
      console.error(`❌ Error exporting results: ${error.message}`);
    }
  }

  // Main analysis method
  analyze() {
    console.log('🚀 Starting CSS Duplicate Analysis...\n');
    
    if (!this.readCSSFile()) {
      return false;
    }

    const rules = this.parseCSS();
    console.log(`📋 Parsed ${rules.length} CSS rules`);

    this.findDuplicateSelectors(rules);
    this.findSimilarRules(rules);
    this.findDuplicateProperties(rules);

    this.generateReport();
    this.exportResults('css-duplicates-report.json');

    return true;
  }
}

// CLI Usage
if (require.main === module) {
  const cssFile = process.argv[2] || 'frontend/src/styles/globals.css';
  
  if (!fs.existsSync(cssFile)) {
    console.error(`❌ CSS file not found: ${cssFile}`);
    console.log('Usage: node css-duplicate-checker.js [path-to-css-file]');
    process.exit(1);
  }

  const checker = new CSSDuplicateChecker(cssFile);
  checker.analyze();
}

module.exports = CSSDuplicateChecker;
